### PR TITLE
JSON date parse helper 📅 ⏰

### PR DESCRIPTION
**Pull request by portellaa**
Created helpers to parse and format a date from a JSON

Type on JSON is inferred from the formatter closure input type.

There is also an Optional version of the API.

### CODE REVIEW

File: Sources/Network/JSON.swift [hunk 1]
```diff
@@ -262,6 +262,62 @@ public enum JSON {
         return try parseOptionalRawRepresentableAttribute(T.self, key: key, json: json, parseAPIError: parseAPIError)
     }
 
+
```
Comment (p4checo) [inline Sources/Network/JSON.swift]: you forgot the specified type variants 😇

File: Sources/Network/JSON.swift [hunk 2]
```diff
@@ -262,6 +262,62 @@ public enum JSON {
...
+    ///   - parseAPIError: The API error parsing closure.
+    /// - Returns: The Date obtained from the formatter.
+    /// - Throws: An error of type `JSON.Error`, or a domain specific `Swift.Error` produced by `parseAPIError`.
+    public static func parseDateAttribute<T>(key: JSON.AttributeKey,
+                                             json: JSON.Dictionary,
+                                             formatter: (T) -> Date?,
+                                             parseAPIError: ParseAPIErrorClosure? = nil) throws -> Date {
+
+        guard let parsedDate = try parseOptionalDateAttribute(key: key,
+                                                          json: json,
```
Comment (p4checo) [inline Sources/Network/JSON.swift]: bad alignment 🚓 💄

File: Sources/Network/JSON.swift [hunk 3]
```diff
@@ -262,6 +262,62 @@ public enum JSON {
...
+    ///   - parseAPIError: The API error parsing closure.
+    /// - Returns: The Date obtained from the formatter or nil.
+    /// - Throws: An error of type `JSON.Error`, or a domain specific `Swift.Error` produced by `parseAPIError`.
+    public static func parseOptionalDateAttribute<T>(key: JSON.AttributeKey,
+                                                     json: JSON.Dictionary,
+                                                     formatter: (T) -> Date?,
+                                                     parseAPIError: ParseAPIErrorClosure? = nil) throws -> Date? {
+
+        guard let rawValue = json[key] else {
+            throw parseAPIError?(json) ?? Error.missingAttribute(key, json: json)
```
Comment (p4checo) [inline Sources/Network/JSON.swift]: this should simply return `nil` since the attribute is optional, or throw the API Error if one is returned by the `parseAPIError` 

like: 
```swift
        guard let rawValue = json[key] else {
            if let apiError = parseAPIError?(json) { throw apiError }
            return nil
        }
```

File: Sources/Network/JSON.swift [hunk 4]
```diff
@@ -262,6 +262,62 @@ public enum JSON {
...
+        }
+
+        let jsonValue: T = try parseValue(rawValue: rawValue, key: key, json: json, parseAPIError: parseAPIError)
+
+        return formatter(jsonValue)
```
Comment (p4checo) [inline Sources/Network/JSON.swift]: What if the formatter fails? Shouldn't we throw an exception? 

e.g.: `throw Error.unexpectedAttributeValue(key, json: json)`

File: Sources/Network/JSON.swift [hunk 5]
```diff
@@ -262,6 +262,62 @@ public enum JSON {
...
+    /// an attempt is made to extract a domain specific error and throw it.
+    ///
+    /// - Parameters:
+    ///   - key: The JSON attribute key to parse.
+    ///   - json: The JSON dictionary.
+    ///   - formatter: The formatter to convert the parsed type into Date.
+    ///   - parseAPIError: The API error parsing closure.
+    /// - Returns: The Date obtained from the formatter.
+    /// - Throws: An error of type `JSON.Error`, or a domain specific `Swift.Error` produced by `parseAPIError`.
+    public static func parseDateAttribute<T>(key: JSON.AttributeKey,
```
Comment (p4checo) [inline Sources/Network/JSON.swift]: on these API variants the `key` parameter is omitted `_ key:` 😉

File: Tests/AlicerceTests/Network/JSONTests.swift [hunk 6]
```diff
@@ -1531,6 +1531,121 @@ class JSONTests: XCTestCase {
...
+        }(DateFormatter())
+
+        let json = [
+            "testDate" : dateFormatter.string(from: testDate)
+        ]
+
+        do {
+            let date = try JSON.parseDateAttribute(key: "testDate", json: json, formatter: dateFormatter.date)
+
+            XCTAssertEqual(testDate, date)
```
Comment (p4checo) [inline Tests/AlicerceTests/Network/JSONTests.swift]: we could assert `assertDumpsEqual` on the `json`s 😛

File: Tests/AlicerceTests/Network/JSONTests.swift [hunk 7]
```diff
@@ -1531,6 +1531,121 @@ class JSONTests: XCTestCase {
...
+            let date = try JSON.parseOptionalDateAttribute(key: "testDate",
+                                                           json: json,
+                                                           formatter: Date.init(timeIntervalSince1970:))
+
+            XCTAssertNil(date)
+        } catch {
+            XCTFail("🔥: unexpected error \(error)")
+        }
+    }
+
```
Comment (p4checo) [inline Tests/AlicerceTests/Network/JSONTests.swift]: There's a missing test where the raw value is right, but the conversion fails with an error (which currently returns `nil`)

File: Sources/Network/JSON.swift [hunk 8]
```diff
@@ -304,13 +307,14 @@ public enum JSON {
...
                                                      formatter: (T) -> Date?,
                                                      parseAPIError: ParseAPIErrorClosure? = nil) throws -> Date? {
 
         guard let rawValue = json[key] else {
-            throw parseAPIError?(json) ?? Error.missingAttribute(key, json: json)
+            if let apiError = parseAPIError?(json) { throw apiError }
+            return nil
         }
 
         let jsonValue: T = try parseValue(rawValue: rawValue, key: key, json: json, parseAPIError: parseAPIError)
```
Comment (p4checo) [inline Sources/Network/JSON.swift]: missing the above check here on the `formatter()` call too 😉

File: Sources/Network/JSON.swift (hunk 4 above)
Comment (portellaa) [reply Sources/Network/JSON.swift]: But this is the case where it can be Optional, right? so the formatter may not be able to convert it, fair enough, but it returns nil. 
it doesn't seems wrong to me

File: Sources/Network/JSON.swift (hunk 8 above)
Comment (portellaa) [reply Sources/Network/JSON.swift]: what do you mean? should we call the `apiError` when the formatter fails?

File: Tests/AlicerceTests/Network/JSONTests.swift (hunk 7 above)
Comment (portellaa) [reply Tests/AlicerceTests/Network/JSONTests.swift]: this one 👉 `testParseDateAttribute_WithInvalidStringForFormatter_ShouldFailWithUnexpectedAttributeValue`

File: Sources/Network/JSON.swift (hunk 8 above)
Comment (p4checo) [reply Sources/Network/JSON.swift]: From my perspective, the `parseOptional` should only return `nil` if the key doesn't come in the JSON at all, and not in any other case.

So, if the key is present in the JSON, we should treat it as a normal `parseAttribute`, and in this case if the `formatter()` call fails and returns `nil`, it means that the data was invalid and we should notify the caller. Makes sense? 😛

File: Tests/AlicerceTests/Network/JSONTests.swift (hunk 7 above)
Comment (p4checo) [reply Tests/AlicerceTests/Network/JSONTests.swift]: Yes, I think so 😉
Also, we should do the same for the `parseOptionalDateAttribute` variant 😇

Comment (p4checo) [reply Tests/AlicerceTests/Network/JSONTests.swift]: Yes, it is 😉 
Still, we should do the same test for the `parseOptionalDateAttribute` variant 😇

File: Sources/Network/JSON.swift (hunk 4 above)
Comment (p4checo) [reply Sources/Network/JSON.swift]: From my understanding, the `parseOptionalAttribute` should only return `nil` if the JSON doesn't contain a value for that key. If it does contain the value, it should essentially behave like a normal `parseAttribute`.

So in this case, I think we should preserve the behavior from the other one in regards to invoking `formatter()` 😜

Comment (portellaa) [reply Sources/Network/JSON.swift]: agree ✅

File: Tests/AlicerceTests/Network/JSONTests.swift [hunk 9]
```diff
@@ -1531,6 +1531,155 @@ class JSONTests: XCTestCase {
...
+        let json = [
+            "testDate" : testDate.timeIntervalSince1970
+        ]
+
+        do {
+            let date = try JSON.parseDateAttribute("testDate",
+                                                   json: json,
+                                                   formatter: Date.init(timeIntervalSince1970:))
+
+            XCTAssertEqual(date.timeIntervalSince1970, testDate.timeIntervalSince1970)
```
Comment (p4checo) [inline Tests/AlicerceTests/Network/JSONTests.swift]: I know it should be equivalent, but shouldn't we compare `date` and `testDate` directly? 💅

File: Tests/AlicerceTests/Network/JSONTests.swift [hunk 10]
```diff
@@ -1531,6 +1531,155 @@ class JSONTests: XCTestCase {
...
+            "testDate" : testDate.timeIntervalSince1970
+        ]
+
+        do {
+            let date = try JSON.parseOptionalDateAttribute("testDate",
+                                                           json: json,
+                                                           formatter: Date.init(timeIntervalSince1970:))
+
+            XCTAssertNotNil(date)
+            XCTAssertEqual(date?.timeIntervalSince1970, testDate.timeIntervalSince1970)
```
Comment (p4checo) [inline Tests/AlicerceTests/Network/JSONTests.swift]: same as above

File: Tests/AlicerceTests/Network/JSONTests.swift (hunk 9 above)
Comment (portellaa) [reply Tests/AlicerceTests/Network/JSONTests.swift]: yeeaaahhhhh it makes sense 🤔 ok

**Comment by portellaa** [comment]
I tried this, but it fails, even if the printed date is the same 😳 so i will leave this with the `timeIntervalSince1970` is that ok? @p4checo

**Comment by p4checo** [comment]
Fine by me 👌😉
🚀
